### PR TITLE
extend CasperTest

### PR DESCRIPTION
**Pull request by rodriguez-facundo**
this requires latest metadata branch in netpyne repo

updating casperTest to explore, populate, check and delete rules...

#### the steps for each card are:

- **Create**: create 2 rules.
- **Populate:** select 1st rule and go through all fields inserting values.
- **Check:** move to the 2nd rule and check it is empty. Then come back to rule 1 and check that the fields are the ones used in step **Populate.**
- **Rename:** delete rule 2 and rename rule 1. Then check all fields again
- **leave:** collapse card.

I started doing something customized for each rule, but maybe it would be a good idea to have a common-ground for all cards. 

And then create a Trello card to discuss the best way to customize the test for each card.

### TEST:

- Run casper test: `casperjs test netpyne-tests.js --host=http://localhost:8888/ --engine=slimerjs`

### CODE REVIEW

**Comment by adrianq** [comment]
Hey @rodriguez-facundo @ddelpiano @tarelli is this working for any of you? I am aware this is passing in [travis](https://travis-ci.org/MetaCell/NetPyNE-UI/jobs/412865138) but it is not working for me...
I am getting an error inside the [rename rule method](https://github.com/MetaCell/geppetto-netpyne/blob/casper-test-rangeComponent/tests/netpyne-tests.js#L2264) which is called from multiple places. Error is not particularly verbose: `FAIL TypeError: domWindowUtils.sendKeyEvent is not a function ` and couldn't find much at Internet. If tests are actually passing for any of you, I am happy to review the code and trust travis from now on. Thanks!

**Comment by rodriguez-facundo** [comment]
it is working for me with no error.  Are you sync with netpyne/metadata_facu and netpyne_ui/casper-test-rangeComponent?

I am seeing that this last branch is closed now. no sure why

**Comment by adrianq** [comment]
btw @rodriguez-facundo this is not WIP anymore, right?

**Comment by rodriguez-facundo** [comment]
yeah. I was not sure this was the right approach to the test, so I left the title in case something different is required.

**Comment by adrianq** [comment]
Yep, I am in those branches. Except for netpyne that I am in metadata because I think you already merged your PR a few days ago -> https://github.com/Neurosim-lab/netpyne/pull/346 Rest should be fine I will wait for @ddelpiano or @tarelli to confirm it is working for them

File: NetPyNETabs.js [hunk 1]
```diff
@@ -144,7 +144,7 @@ export default class NetPyNETabs extends React.Component {
           contentContainerStyle={{ bottom: bottomValue, position: 'absolute', top: 48, left: 0, right: 0, overflow: 'auto' }}
           onChange={this.handleChange}
         >
-          <Tab label="Define your network" value="define" id={"defineNetwork"}>
+          <Tab  label="Define your network" value="define" id={"defineNetwork"}>
```
Comment (adrianq) [inline NetPyNETabs.js]: no need for this extra space

File: components/definition/cellRules/NetPyNECellRule.js [hunk 2]
```diff
@@ -77,19 +78,19 @@ export default class NetPyNECellRule extends React.Component {
             <b>Conditions:</b>
           </div>
 
-          <NetPyNEField id={"netParams.cellParams.conds.cellModel"} >
+          <NetPyNEField id={"netParams.cellParams.conds.cellType"} >
```
Comment (adrianq) [inline components/definition/cellRules/NetPyNECellRule.js]: Are you swapping these fields on purpose?

File: components/definition/populations/NetPyNEPopulation.js [hunk 3]
```diff
@@ -15,6 +15,19 @@ import NetPyNEField from '../../general/NetPyNEField';
 import DimensionsComponent from './Dimensions';
 import NetPyNECoordsRange from '../../general/NetPyNECoordsRange';
 
+/*
```
Comment (adrianq) [inline components/definition/populations/NetPyNEPopulation.js]: What is this?

File: components/definition/populations/NetPyNEPopulation.js [hunk 4]
```diff
@@ -124,51 +137,48 @@ export default class NetPyNEPopulation extends React.Component {
...
               model={"netParams.popParams['" + this.props.name + "']['cellType']"}
-              id={"popCellType"}
             />
           </NetPyNEField>
-
+          
           <NetPyNEField id="netParams.popParams.cellModel" >
-            <PythonControlledAutoComplete
-              dataSource={[]}
+            <PythonControlledTextField
```
Comment (adrianq) [inline components/definition/populations/NetPyNEPopulation.js]: Did you do this on purpose? If so, can you expand the PR description with the changes you are doing apart from the casper extension?

File: components/definition/stimulationSources/NetPyNEStimulationSource.js [hunk 5]
```diff
@@ -64,12 +64,18 @@ export default class NetPyNEStimulationSource extends React.Component {
...
-      Utils
-        .sendPythonMessage("'" + value + "' in netParams.stimSourceParams['" + this.state.currentName + "']['type']")
-        .then((response) => { if (response == true) { this.setState({ sourceType: value }) } });
-    };
-    this.stimSourceTypeOptions.forEach((option) => { getType(option.type) });
+    var opts = this.stimSourceTypeOptions.map((option) => { return option.type });
+    Utils
+      .sendPythonMessage("[value == netParams.stimSourceParams['" + this.state.currentName + "']['type'] for value in "+JSON.stringify(opts)+"]")
+      .then((responses) => {
+        if (responses.constructor.name == "Array"){
```
Comment (adrianq) [inline components/definition/stimulationSources/NetPyNEStimulationSource.js]: I am bit lost here... what are you actually changing?

File: tests/netpyne-tests.js [hunk 6]
```diff
@@ -1,505 +1,2565 @@
...
-		test.assertExists('div[id="Connections"]', "Connections button exists.");
-		test.assertExists('div[id="SimulationSources"]', "Connections button exists.");
-		test.assertExists('div[id="Configuration"]', "Configuration button exists.");
-		test.assertExists('button[id="defineNetwork"]', 'Define network button exists.');
-		test.assertExists('button[id="exploreNetwork"]', 'Explore network button exists.');
-		test.assertExists('button[id="simulateNetwork"]', 'Simulate network button exists.');
-	});
+function testLandingPage(test) {
+  casper.then(function() {
+    assertExist(test, "Populations", "div")
```
Comment (adrianq) [inline tests/netpyne-tests.js]: Why did you change this bit?

File: tests/netpyne-tests.js [hunk 7]
```diff
@@ -1,505 +1,2565 @@
...
+  casper.thenClick('button[id="Section 2"]', function() { // change to rule 2
+    this.echo("go to section 2 -> values must be empty")
+    this.wait(2500) // let python populate fields  
+  })
+
+  casper.then(function() { // check values must be empty
+    checkSectionGeomTabValues(test, "CellRule", "Section 2", "", "", "", "", "", "")
+  })
+
+  casper.thenClick('button[id="Section"]', function() { // back to s
```
Comment (adrianq) [inline tests/netpyne-tests.js]: This is more like renameEntity or sth like that, right?

File: components/definition/populations/NetPyNEPopulation.js (hunk 4 above)
Comment (rodriguez-facundo) [reply components/definition/populations/NetPyNEPopulation.js]: yes, this is an important one. The auto-complete in cellModel behaves very weird. You enter a value, and the field goes empty, then the value reappears, then it goes empty again, then it triggers a refresh again, and finally it gets empty. 
If you try to close the card and re-open it, you get the same behavior. It was imposible to assert the value of the fields in the casper-test. I had to change it.

File: components/definition/cellRules/NetPyNECellRule.js (hunk 2 above)
Comment (rodriguez-facundo) [reply components/definition/cellRules/NetPyNECellRule.js]: cellType is a field that the user has to complete in order for netpyne to work, so I thought it would be better to have it first (I should open this in a different PR, but the change was minor so I decided to put it in here)

File: tests/netpyne-tests.js (hunk 6 above)
Comment (rodriguez-facundo) [reply tests/netpyne-tests.js]: sometimes the casperTest fails saying that you cannot click inexistent "Populations" selector.
I standardized the way we assert things exist by first calling `this.waitUntilVisible`and then asserting it exist. Sometimes it was like this but in other parts it was not. So now the same function is use everywhere

File: components/definition/stimulationSources/NetPyNEStimulationSource.js (hunk 5 above)
Comment (rodriguez-facundo) [reply components/definition/stimulationSources/NetPyNEStimulationSource.js]: We have **['VClamp', 'IClamp', 'NetStim',  'AlphaSynapse', 'SEClamp']** as options for this menu. Instead of sending 5 python messages asking if a variable is equal to one of them, I send an unique python message which returns: 
**[False, False, False, True, False]**, so I know the 4th element is the one I need to pick.
The python message code translates into this:

for value in opts:
....if value == netParams.stimSourceParams[blabla][blabla]:
........return True
....else:
........return False

File: tests/netpyne-tests.js (hunk 7 above)
Comment (rodriguez-facundo) [reply tests/netpyne-tests.js]: yeah, but I created a function to rename rules like popParams rules, conParams rules, but because it also works for renaming sections and for renaming mechanisms, I just used it there

**Comment by rodriguez-facundo** [comment]
following @adrianq suggestion, I've divided into separated files to make it easier to read and modify.

File: components/definition/cellRules/NetPyNECellRule.js (hunk 2 above)
Comment (adrianq) [reply components/definition/cellRules/NetPyNECellRule.js]: OK, it is not a big deal but try to do this kind of changes in different branches. It was already a big PR

File: components/definition/populations/NetPyNEPopulation.js (hunk 4 above)
Comment (adrianq) [reply components/definition/populations/NetPyNEPopulation.js]: Ok, we can also try to fix instead of just replacing it. So... aren't we using autocomplete anywhere else in the code?

File: components/definition/stimulationSources/NetPyNEStimulationSource.js (hunk 5 above)
Comment (adrianq) [reply components/definition/stimulationSources/NetPyNEStimulationSource.js]: Ok, makes sense... although it would be good to move it to a different branch ;) No worries this time

File: components/definition/populations/NetPyNEPopulation.js (hunk 4 above)
Comment (rodriguez-facundo) [reply components/definition/populations/NetPyNEPopulation.js]: no, that is the only field that uses autocomplete. I will create a card for that so we don't forget

**Comment by adrianq** [comment]
Thanks for splitting the tests. Much easier to understand and extend them now. Tests are passing and I have tried to review it as much as possible. I am happy to merge it unless @ddelpiano or @tarelli would like to have a look at it as well.
